crates/interpreter/src/exec.rs

```diff
@@ -1055,15 +1069,18 @@ impl<'m> Thread<'m> {
     fn exit_label(mut self) -> ThreadResult<'m> {
         let frame = self.frame();
         let label = frame.labels.pop().unwrap();
+        let popped_values: Vec<Val> =
+            frame.labels_values.drain(frame.labels_values.len() - label.values_cnt ..).collect();
         if frame.labels.is_empty() {
             let frame = self.frames.pop().unwrap();
-            debug_assert_eq!(label.values.len(), frame.arity);
+            debug_assert_eq!(label.values_cnt, frame.arity);
             if self.frames.is_empty() {
-                return ThreadResult::Done(label.values);
+                return ThreadResult::Done(popped_values);
             }
             unsafe { self.parser.restore(frame.ret) };
         }
-        self.values().extend_from_slice(&label.values);
+        self.values().extend_from_slice(&popped_values);
```

Comment:
This looks again like a case that can be optimized by not draining and not extending. It might be a bit more complicated due to the case where we're exiting the last label of a frame and thus exiting the frame.